core/src/apps/cardano/layout.py

```diff
@@ -104,13 +248,69 @@ async def confirm_sending_token(
     )
 
 
-async def show_warning_tx_output_contains_tokens(ctx: wire.Context) -> None:
-    await confirm_metadata(
+async def show_credential(
+    ctx: wire.Context,
+    credential_params: CredentialParams,
+    credential_policy: int,
+) -> None:
+    props: list[PropertyType] = []
+
+    credential = credential_params.get_credential()
+    # Credential can be empty in case of enterprise address stake credential
+    # and reward address payment credential. In that case we don't want to
+    # show some of the "props".
+    if credential:
+        if (credential_policy & CREDENTIAL_POLICY_SHOW_CHANGE) != 0:
+            address_usage = "Change address"
+        elif (credential_policy & CREDENTIAL_POLICY_SHOW_OUTPUT) != 0:
+            address_usage = "Output address"
+        else:
+            address_usage = "Address"
+
+        credential_title = credential_params.get_title()
+        props.append(
+            (
+                "%s %s credential is a %s:"
+                % (address_usage, credential_params.format_type(), credential_title),
+                None,
+            )
+        )
+        props.extend(credential_params.format())
+
+    if (credential_policy & CREDENTIAL_POLICY_FAIL_OR_WARN_UNUSUAL) != 0:
+        props.append((None, "Path is unusual."))
+    if (credential_policy & CREDENTIAL_POLICY_WARN_MISMATCH) != 0:
+        props.append((None, "Credential doesn't match payment credential."))
+    if (credential_policy & CREDENTIAL_POLICY_WARN_REWARD_ADDRESS) != 0:
+        props.append(("Address is a reward address.", None))
+    if (credential_policy & CREDENTIAL_POLICY_WARN_NO_STAKING) != 0:
+        props.append(
+            (
+                "%s address - no staking rewards."
+                % ADDRESS_TYPE_NAMES[credential_params.address_type],
+                None,
+            )
+        )
+
+    if credential_policy in CREDENTIAL_POLICIES_WARN:
+        icon = ui.ICON_WRONG
+        icon_color = ui.RED
+    else:
+        icon = ui.ICON_SEND
+        icon_color = ui.GREEN
+
+    if (credential_policy & CREDENTIAL_POLICY_SHOW_OUTPUT) != 0:
+        title = "Confirm transaction"
+    else:
+        title = "%s address" % ADDRESS_TYPE_NAMES[credential_params.address_type]
```

Comment:
This perhaps deserves a comment (about this being the place where we distinguish between displaying address derivation and tx output). Maybe it's worth moving to the top of the function.

Reply:
I've moved it to the top of the function. IMHO the comment is not needed now. https://github.com/vacuumlabs/trezor-firmware/pull/66/commits/4d77f4e42747348ce6de09ab62a8e5ae49c07e53#diff-718c35fab848e1ee6f5e2a387c71b0a0ea8d84039fedce530cb9f0584c164495R255